graphql_compiler/compiler/emit_sql.py

```diff
@@ -103,6 +113,112 @@ def _find_folded_outputs(ir):
+    regardless of what characters they contain, whether they are empty or null, and
+    regardless of their data type.
+
+    Empty arrays are represented as '', null values are represented as '~', array
+    entries are delimited using '|'.
+
+    All occurrences of '^', '~', and '|' in the original string values are
+    replaced with '^e' (escape), '^n' (null), and '^d' (delimiter), resp.
+
+    Undoing the encoding above as well as the XML reference entity encoding performed
+    by the XML PATH statement is deferred to post-processing when the list is retrieved
+    from the string representation produced by the subquery.
+
+    Post-processing must split on '|', convert '~' to None, and undo both the encoding above
+    and the XML reference entity encoding auto performed by XML PATH. In particular, undo
+    '^d' -> '|', '^e' -> '^', '^n' -> '~', '&amp;' -> '&', '&gt;' -> '>', '&lt;' -> '<',
+    '&0xHEX;' -> u'0xHEX'. XML post processing can be done with
+    https://docs.python.org/3/library/html.html#html.unescape
+
+    Joining from the preceding vertex to the correct output vertex
+    is performed with the WHERE clause. This is contrasted with PostgreSQL subqueries,
+    in which the join from penultimate vertex to output vertex is done within
+    the FROM/JOIN clause of the aggregation subquery.
+
+    Args:
+        output_column: SQLAlchemy Column, value being aggregated with XML PATH
+        where: SQLAlchemy BinaryExpression, predicate used to filter
+
+    Returns:
+        A SQLAlchemy Selectable corresponding to the XML PATH subquery
+    """
+    encoded_column = func.REPLACE(  # replace all occurrences of '^' in the original with '^e'
+        output_column, expression.literal_column('\'^\''), expression.literal_column('\'^e\'')
+    )
+    encoded_column = func.REPLACE(  # replace all occurrences of '~' in the original with '^n'
+        encoded_column, expression.literal_column('\'~\''), expression.literal_column('\'^n\'')
+    )
+    encoded_column = func.REPLACE(  # replace all occurrences of '|' in the original with '^d'
+        encoded_column, expression.literal_column('\'|\''), expression.literal_column('\'^d\'')
+    )
+
+    # delimit elements in the array using '|'and mark nulls with `~`
+    xml_column = (expression.literal_column('\'|\'') + func.COALESCE(  # denote null values with '~'
+        encoded_column, expression.literal_column('\'~\'')
+    ))  # allow unambiguously distinguishing (nullable) array elements using scheme above
+
+    # use constructor: you can't directly construct an XMLPathBinaryExpression from plain text
+    xml_column = XMLPathBinaryExpression(xml_column.left, xml_column.right, xml_column.operator)
+
+    return func.COALESCE(  # coalesce to represent empty arrays as ''
+        select([xml_column]).where(where).suffix_with("FOR XML PATH ('')").as_scalar(),
+        expression.literal_column('\'\'')
+    )
```

Comment:
We absolutely need integration tests to see if this logic works as expected. You're free to add a new district of vertices in the integration testing graph filled with string valuies with weird characters. Doesn't have to be in this PR.

My advice actually is to split this PR into 2 PRs:
- xml_aggregation: This has nothing to do with `@fold` and can even be its own module. It's just some utils that help you mimic `array_agg` using `xml`. Then you can create some very simple sql queries and check their correctness with integration tests.
- `@fold` for old mssql versions: This would be this PR minus all the xml stuff. Hoever, now the `emit_sql` module is much smaller and easier to understand, and also reviewers already trust that the xml logic works and they can focus their doubt on the other parts of the PR.

In general, try to split your work into small PRs whenever possible. You will have fewer rounds of code reviews and things will go much faster. My method is to get everything working on my own secret WIP branch, make a merge plan and progressively roll out in multiple PRs. For example if you look at my WIP PR https://github.com/kensho-technologies/graphql-compiler/pull/638 you'll see that it's ugly code (that passes tests) with a merge plan that I will execute in 5 PRs. You don't have to follow my method but that's been working well for me.

Reply:
I have factored out the pure fold changes to #644. Please review it when you get the chance.

The MSSQL stuff relies on those changes so I am going to keep this PR open for now.